src/olympia/devhub/views.py

```diff
@@ -2074,3 +2079,99 @@ def logout(request):
+    VERIFY_EMAIL_STATE['verification_expired'],
+    VERIFY_EMAIL_STATE['verification_failed'],
+    VERIFY_EMAIL_STATE['verification_timedout'],
+    VERIFY_EMAIL_STATE['confirmation_unauthorized'],
+]
+
+
+def get_button_text(state):
+    if state == VERIFY_EMAIL_STATE['email_suppressed']:
+        return gettext('Verify email')
+
+    return gettext('Try again')
+
+
+@login_required
+def email_verification(request):
+    data = {'state': None}
+    email_verification = request.user.email_verification
+    suppressed_email = request.user.suppressed_email
+
+    if request.method == 'GET':
+        if email_verification:
+            if not email_verification.is_expired:
+                if email_verification.status_pending:
+                    if email_verification.is_timedout:
+                        data['state'] = VERIFY_EMAIL_STATE['verification_timedout']
+                    else:
+                        data['state'] = VERIFY_EMAIL_STATE['verification_pending']
+                elif email_verification.status_failed:
+                    data['state'] = VERIFY_EMAIL_STATE['verification_failed']
+                elif email_verification.status_delivered:
+                    code = request.GET.get('code')
+
+                    if code:
+                        if code == email_verification.confirmation_code:
+                            suppressed_email.delete()
+                            send_mail_jinja(
+                                gettext('Your email has been verified'),
+                                'devhub/emails/verify-email-completed.ltxt',
+                                {},
+                                recipient_list=[request.user.email],
+                            )
+
+                        if SuppressedEmailVerification.objects.filter(
+                            confirmation_code=code
+                        ).exists():
+                            data['state'] = VERIFY_EMAIL_STATE[
+                                'confirmation_unauthorized'
+                            ]
+                        else:
+                            return redirect('devhub.email_verification')
+                    else:
+                        data['state'] = VERIFY_EMAIL_STATE['confirmation_pending']
+            else:
+                data['state'] = VERIFY_EMAIL_STATE['verification_expired']
+
+        elif suppressed_email:
+            data['state'] = VERIFY_EMAIL_STATE['email_suppressed']
+        else:
+            data['state'] = VERIFY_EMAIL_STATE['email_verified']
+
+    elif request.method == 'POST':
+        if email_verification:
+            email_verification.delete()
+
+        if suppressed_email:
+            email_verification = SuppressedEmailVerification.objects.create(
+                suppressed_email=suppressed_email
+            )
+            send_suppressed_email_confirmation.delay(email_verification.id)
+
+        return redirect('devhub.email_verification')
```

Comment:
maybe move this above the GET because it returns in all cases so the GET case can be dedented.

Reply:
Fair enough.